modules/33acrossAnalyticsAdapter.js

```diff
@@ -0,0 +1,463 @@
+
+  return DEFAULT_TRANSACTION_TIMEOUT;
+}
+
+/**
+ * @param {TransacionManager} transactionManager
+ */
+function subscribeToGamSlotRenderEvent(transactionManager) {
+  window.googletag = window.googletag || {};
+  window.googletag.cmd = window.googletag.cmd || [];
+  window.googletag.cmd.push(() => {
+    window.googletag.pubads().addEventListener('slotRenderEnded', event => {
+      log.info('slotRenderEnded', `adUnitPath: ${event.slot.getAdUnitPath()}`, `slotElementId: ${event.slot.getSlotElementId()}`);
+    });
+  });
+}
+
+/** necessary for testing */
+analyticsAdapter.originDisableAnalytics = analyticsAdapter.disableAnalytics;
+analyticsAdapter.disableAnalytics = function () {
+  analyticsAdapter._oldEnable = enableAnalyticsWrapper;
+  locals.reset();
+  analyticsAdapter.originDisableAnalytics();
+};
+
+adapterManager.registerAnalyticsAdapter({
+  adapter: analyticsAdapter,
+  code: PROVIDER_NAME,
+  gvlid: 58,
+});
+
+export default analyticsAdapter;
+
+/**
+ * @param {string} pid Partner ID
+ * @returns {AnalyticsCache}
+ */
+function newAnalyticsCache(pid) {
+  return {
+    pid,
+    auctions: {},
+    bidsWon: {},
+  };
+}
+
+/**
+ * @param {AnalyticsCache} analyticsCache
+ * @param {string} completedAuctionId value of auctionId
+ * @return {AnalyticsReport} Analytics report
+ */
+function createReportFromCache(analyticsCache, completedAuctionId) {
+  const { pid, bidsWon, auctions } = analyticsCache;
+
+  return {
+    pid,
+    src: 'pbjs',
+    analyticsVersion: ANALYTICS_VERSION,
+    pbjsVersion: '$prebid.version$', // Replaced by build script
+    auctions: [ auctions[completedAuctionId] ],
+    bidsWon: bidsWon[completedAuctionId]
+  }
+}
+
+/**
+ * @param {Object} args
+ * @param {Array} args.adUnits
+ * @param {string} args.auctionId
+ * @param {Array} args.bidderRequests
+ * @returns {Auction}
+ */
+function parseAuction({ adUnits, auctionId, bidderRequests }) {
+  if (typeof auctionId !== 'string' || !Array.isArray(bidderRequests)) {
+    log.error('Analytics adapter failed to parse auction.');
+  }
+
+  return {
+    adUnits: adUnits.map(unit => parseAdUnit(unit)),
+    auctionId,
+    userIds: Object.keys(utils.deepAccess(bidderRequests, '0.bids.0.userId', {}))
+  }
+}
+
+/**
+ * @param {Object} args
+ * @param {string} args.transactionId
+ * @param {string} args.code
+ * @param {string} args.ortb2Imp
+ * @param {Array<string>} args.mediaTypes
+ * @param {Array<string>} args.sizes
+ * @returns {AdUnit}
+ */
+function parseAdUnit({ transactionId, code, ortb2Imp, mediaTypes, sizes }) {
+  return {
+    transactionId,
+    adUnitCode: code,
+    // Note: GPID supports adUnits that have matching `code` values by appending a `#UNIQUIFIER`.
+    // The value of the UNIQUIFIER is likely to be the div-id,
+    // but, if div-id is randomized / unavailable, may be something else like the media size)
+    slotId: utils.deepAccess(ortb2Imp, 'ext.gpid', code),
```

Comment:
just want to understand what values we expect on this field.. I am seeing a value like ` "slotId": "/17118521/prebid_adunit1"
` in the payload document, so it looks like the value that could be obtained from GAM when we call `event.slot.getAdUnitPath()`

Reply:
They will often be the same, but they are not always. There are a number of different permutations that can result in the slotId, depending on how the publisher defined the slot. GPID works as a source of truth in these scenarios, and is populated as long as gptPreAuction is compiled into prebid OR if the publisher includes a gpid outright in the specification of the adUnit. The goal of the former is to have slotIds that are consistent across loads even if some aspects of the slot are dynamic (whether that be div-id, adUnit code, or whether the ad slots have duplicate names).

GPID is populated at BEFORE_REQUEST_BIDS.

See more [here](https://docs.prebid.org/features/pbAdSlot.html) and [here](https://docs.prebid.org/dev-docs/modules/gpt-pre-auction.html)

Reply:
do we know how common is the inclusion of gptPreAuction when the publisher has a prebid.js+GAM integration?

I like the idea behind GPID, just wondering if we should populate the slotId field with something else when GPID field is not present.

Reply:
For publishers not using GAM, the prebid.js documentation specifies that publishers are *required* to supply the value manually on the adUnit. But, I've added a fallback to `ortb2Imp.ext.data.pbadslot`, and another fallback from there to instead use the adUnitCode, if those are missing.